2단계 - 문자열 덧셈 계산기 실습 PR 요청드립니다!

taewonh opened the pull request.

taewonh committed:
test(StringAddCalculatorTest): 테스트 코드 추가

- 테스트 코드의 컴파일 오류를 제거하기 위해 동작하지 않는 StringAddCalculator 클래스 및 splitAndSum 메소드 추가

taewonh committed:
feat(StringAddCalculator):  기능 구현

heowc commented on src/main/java/StringAddCalculator.java:
```diff
@@ -0,0 +1,62 @@
+    private static final Pattern delimiterPattern = Pattern.compile("//(.)\n(.*)");
+
+    public static int splitAndSum(String text) {
+        if (isEmpty(text)) {
```
꼼꼼한 문자열 체크 좋습니다 👍

heowc commented on src/main/java/StringAddCalculator.java:
```diff
@@ -0,0 +1,62 @@
+            return 0;
+        }
+        Matcher delimiterMatcher = delimiterPattern.matcher(text);
+        text = removeDelimiter(text, delimiterMatcher);
+        String delimiter = generateDelimiter(delimiterMatcher);
```
메소드 순서를 호출 순서대로 나열해보는 것은 어떨까요? 🤔

taewonh replied:
@heowc 메소드 선언 순서와 동일하게 호출 순서를 변경하였습니다!

taewonh replied:
@heowc 별개로 커밋하다보니 커밋 메시지에 fix로 입력 했는데 의미상 refactoring이 좀더 맞을까요..?

heowc replied:
리팩토링이 더 적절해보이긴 하네요.
https://gist.github.com/stephenparish/9941e89d80e2bc58a153를 참고하시면 더 도움이 될 것 같습니다 👍

heowc commented on src/main/java/StringAddCalculator.java:
```diff
@@ -0,0 +1,62 @@
+    private static int sum(String[] split) {
+        return Arrays.stream(split)
+                .mapToInt(StringAddCalculator::parseInt)
+                .peek(StringAddCalculator::checkMinus)
```
`peek(...)`으로 구현한 것은 처음 보는 것 같습니다. 좋네요 👍

heowc commented on src/main/java/StringAddCalculator.java:
```diff
@@ -0,0 +1,62 @@
+        try {
+            return Integer.parseInt(text);
+        } catch (Exception e) {
+            throw new RuntimeException();
```
파싱하는 부분과 음수인지 체크하는 부분 잘 나눠주셨습니다 👍 

다만, `RuntimeException`으로만 예외를 던지기 보다는 상황에 맞는 예외를 만들어 던지는 것은 어떨까요? 🤔 
추가로 해당 라인에 대한 테스트도 추가해주시면 좋을 것 같아요 🙇

taewonh replied:
@heowc RuntimeException Throw 로 적용된 메소드 2가지 (parseInt, checkMinus) 모두 의미상 좀더 부합하도록 IllegalArgumentException으로 지정하고, 오류 메세지도 구분해서 입력하도록 처리했습니다! 추가로 두 가지 케이스 모두 검증가능하도록 기존 테스트 코드 역시 2가지로 분리했습니다!

heowc reviewed (changes requested):
안녕하세요 태원님! 🙇 
문자열 덧셈 계산기 잘 구현해주셨습니다. 👍 

코멘트 몇 개 남겼으니 확인 부탁드려요!

궁금하신 점이나 이야기 나누고 싶은 부분이 있다면 깃헙 코멘트, 슬랙 DM 편하게 남겨주세요!

taewonh committed:
fix(StringAddCalculator): 메소드 호출 순서 변경

- 메소드 선언과 동일하게 호출 순서 변경

taewonh committed:
refactor(StringAddCalculator): parseInt, checkMinus 예외처리 클래스 변경, 테스트 분리

- RuntimeException이 아닌 의미상 조금 더 부합하는 IllegalArgumentException으로 변경
- Exception Throw 시 상황에 맞는 메시지 입력
- parseInt, checkMinus 예외 테스트 분리 (asis : splitAndSum_negative)

heowc reviewed (approved):
크게 피드벡 드릴게 없어보이네요.
다음 단계를 진행해주세요 🙇 

다음 미션부터 본격적인 미션이 시작되니 화이팅 입니다 👍
